### DISCUSSION TITLE
How to create MeshShape

### DISCUSSION THREAD

**Discussion by kutu** [jrouwe/JoltPhysics.js#69] [Q&A] [answered]
Following example `MeshShapeTest.cpp`

I have ThreeJS Mesh geometry, I use it to create MeshShapeSettings

```js
g = this.mesh.geometry.getAttribute('position');
gi = this.mesh.geometry.index;

tList = new Jolt.TriangleList();

for (i = j = 0, ref = gi.count; (0 <= ref ? j < ref : j > ref); i = 0 <= ref ? ++j : --j) {
  tList.push_back(new Jolt.Triangle(
 new Jolt.Vec3(g.getX(gi.getX(i)), g.getY(gi.getX(i)), g.getZ(gi.getX(i))),
 new Jolt.Vec3(g.getX(gi.getY(i)), g.getY(gi.getY(i)), g.getZ(gi.getY(i))),
 new Jolt.Vec3(g.getX(gi.getZ(i)), g.getY(gi.getZ(i)), g.getZ(gi.getZ(i)))
));
}

body = bodyInterface.CreateBody(new Jolt.BodyCreationSettings(new Jolt.MeshShapeSettings(tList), new Jolt.Vec3(), new Jolt.Quat(), Jolt.EMotionType_Static, LAYER_NON_MOVING));
bodyInterface.AddBody(body.GetID(), Jolt.EActivation_Activate);
```

Then I tried to use vertices+indexedTriangles

```js
g = this.mesh.geometry.getAttribute('position');

vList = new Jolt.VertexList();

for (i = j = 0, ref = g.count; (0 <= ref ? j < ref : j > ref); i = 0 <= ref ? ++j : --j) {
  vList.push_back(new Jolt.Float3(g.getX(i), g.getY(i), g.getZ(i)));
}

gi = this.mesh.geometry.index;
tList = new Jolt.IndexedTriangleList();

for (i = k = 0, ref1 = gi.count; (0 <= ref1 ? k < ref1 : k > ref1); i = 0 <= ref1 ? ++k : --k) {
  // this throws error
  // tList.push_back(new Jolt.IndexedTriangle(gi.getX(i), gi.getY(i), gi.getZ(i), 0));
  it = new Jolt.IndexedTriangle();
  it.set_mIdx(0, gi.getX(i));
  it.set_mIdx(1, gi.getY(i));
  it.set_mIdx(2, gi.getZ(i));
  tList.push_back(it);
}

mList = new Jolt.PhysicsMaterialList();
mList.push_back(new Jolt.PhysicsMaterial());

body = bodyInterface.CreateBody(new Jolt.BodyCreationSettings(new Jolt.MeshShapeSettings(vList, tList, mList), new Jolt.Vec3(), new Jolt.Quat(), Jolt.EMotionType_Static, LAYER_NON_MOVING));
bodyInterface.AddBody(body.GetID(), Jolt.EActivation_Activate);
```

In both cases I get the same error
![image](https://github.com/jrouwe/JoltPhysics.js/assets/228802/a79b739d-8f36-4522-84a9-025c9e7188a3)

Error for commented line indexed triangle creation
![image](https://github.com/jrouwe/JoltPhysics.js/assets/228802/ed8f5b55-b9c4-4c35-9e7f-fd1137686ae6)

And there was no `PhysicsMaterialSimple` so I used `PhysicsMaterial` (don't know if it matters). I have to use material list, because in 2nd signature function, material list parameter is not optional, like in 1st one.
```js
    class MeshShapeSettings extends ShapeSettings {
        constructor(inTriangleList: TriangleList, inMaterialList?: PhysicsMaterialList);
        constructor(inVertices: VertexList, inTriangles: IndexedTriangleList, inMaterialList: PhysicsMaterialList);
    }
```

**Comment by LeXXik** [comment]
Speaking of which, I am curious if we could pass vertex and index typed arrays/buffers directly to avoid copying data. I know that wasm_bindgen supports it, but not sure about Emscripten.

**Comment by jrouwe** [comment, marked as answer]
> body = bodyInterface.CreateBody(new Jolt.BodyCreationSettings(new Jolt.MeshShapeSettings(tList), new Jolt.Vec3(), new Jolt.Quat(), Jolt.EMotionType_Static, LAYER_NON_MOVING));

There are a number of things wrong with this line:
1. The first parameter is a Shape, not a ShapeSettings object, so ```new Jolt.MeshShapeSettings(tList)``` should be ```new Jolt.MeshShapeSettings(tList).Create().Get()```
2. ```new Jolt.Vec3()``` creates an uninitialized vector, you need to use ```Jolt.Vec3.prototype.sZero()``` or ```new Jolt.Vec3(0, 0, 0)```
3. Same for ```new Jolt.Quat()```, you need to use ```Jolt.Quat.prototype.sIdentity()``` or ```new Jolt.Quat(0, 0, 0, 1)```

> // this throws error
> // tList.push_back(new Jolt.IndexedTriangle(gi.getX(i), gi.getY(i), gi.getZ(i), 0));

This just seems to work for me (but maybe you can give me a full example that I can just load directly to repro).

> And there was no PhysicsMaterialSimple so I used PhysicsMaterial (don't know if it matters).

That's fine. In C++ you inherit from PhysicsMaterial to add additional custom properties, but in JavaScript you can add any property you like so no need for inheritance.

> I have to use material list, because in 2nd signature function, material list parameter is not optional, like in 1st one.

Yes, this is an unfortunate side effect of the way WebIDL/Emscripten works. In C++ you can create different overloads for the same function as long as the types are clear, but in JavaScript you have to differentiate between the number of parameters. So if I make both functions have an optional PhysicsMaterialList then the 2nd function complains:

```overloading must differentiate by # of arguments (cannot have two signatures that differ by types but not by length)```

> Speaking of which, I am curious if we could pass vertex and index typed arrays/buffers directly to avoid copying data. I know that wasm_bindgen supports it, but not sure about Emscripten.

On the C++ side this class uses a move operation, so it normally doesn't copy the data. How that works with JavaScript I'm not sure. What I can do is to expose the default constructor and expose the internal members so you can directly add the data there.

**Reply by LeXXik** [reply]
> On the C++ side this class uses a move operation, so it normally doesn't copy the data. How that works with JavaScript I'm not sure. What I can do is to expose the default constructor and expose the internal members so you can directly add the data there.

Hmm, not sure how not to copy at the moment. The way I do it at the moment is that I iterate over an index buffer of a visual mesh to populate a triangle list of the collider shape. Which I guess is an equivalent of a copy. For example, in Rapier physics engine, which is based on Rust wasm_bindgen, we create a trimesh using `new Rapier.TriMesh(vertices: Float32Array, indices: Uint32Array);`. We can just pass the visual mesh buffer views to it to generate a trimesh. While in Ammo (Bullet), we do it also by iterating over indices/vertices to generate triangles to generate a final trimesh. Similarly, like I do here in Jolt.

**Reply by kutu** [reply]
Thank you again.

`.Create().Get()` this fixed the problem.

Creating `IndexedTriangle` is actually working, problem was on THREE side, the `index` buffer attribute `count` variable provides just total number of indexes, not divided by 3 (so again my fault)

`new Jolt.Vec3()` does automatically set all components to `0` https://jsfiddle.net/af30o8qe/

`Quat` also sets all to `0`, so yes, we need to set `W` to `1` manually or use static function (but it somehow works anyway with all `0`). But like in THREE, other devs would expect default values for `new Jolt.Quat()`.

**Reply by LeXXik** [reply]
> new Jolt.Vec3() does automatically set all components to 0 https://jsfiddle.net/af30o8qe/

I'm afraid it is not correct. When a vector is initialized without parameters, the components will be uninitialized. In this case their values will depend on what was happening in the Wasm memory immediately before the initialization of the vector. Often it will be zero, but can be some random number. I usually use `new Jolt.Vec3(0, 0, 0);` and `new Jolt.Quat(0, 0, 0, 1);`, if I need them to have default values.

**Reply by kutu** [reply]
Good to know, thanks for info.
